README.rst
..
    This file is part of Invenio.
    Copyright (C) 2016 RERO.

    Invenio is free software; you can redistribute it
    and/or modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation; either version 2 of the
    License, or (at your option) any later version.

    Invenio is distributed in the hope that it will be
    useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Invenio; if not, write to the
    Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
    MA 02111-1307, USA.

    In applying this license, RERO does not
    waive the privileges and immunities granted to it by virtue of its status
    as an Intergovernmental Organization or submit itself to any jurisdiction.

==============================
 Invenio-Collections-Metadata
==============================

.. image:: https://img.shields.io/travis/jma/invenio-collections-metadata.svg
        :target: https://travis-ci.org/jma/invenio-collections-metadata

.. image:: https://img.shields.io/coveralls/jma/invenio-collections-metadata.svg
        :target: https://coveralls.io/r/jma/invenio-collections-metadata

.. image:: https://img.shields.io/github/tag/jma/invenio-collections-metadata.svg
        :target: https://github.com/jma/invenio-collections-metadata/releases

.. image:: https://img.shields.io/pypi/dm/invenio-collections-metadata.svg
        :target: https://pypi.python.org/pypi/invenio-collections-metadata

.. image:: https://img.shields.io/github/license/jma/invenio-collections-metadata.svg
        :target: https://github.com/jma/invenio-collections-metadata/blob/master/LICENSE


Add metadata to Invenio collections.

*This is an experimental developer preview release.*

* Free software: GPLv2 license
* Documentation: https://pythonhosted.org/invenio-collections-metadata/


CHANGES.rst
..
    This file is part of Invenio.
    Copyright (C) 2016 RERO.

    Invenio is free software; you can redistribute it
    and/or modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation; either version 2 of the
    License, or (at your option) any later version.

    Invenio is distributed in the hope that it will be
    useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Invenio; if not, write to the
    Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
    MA 02111-1307, USA.

    In applying this license, RERO does not
    waive the privileges and immunities granted to it by virtue of its status
    as an Intergovernmental Organization or submit itself to any jurisdiction.


Changes
=======

Version 0.1.0 (released TBD)

- Initial public release.


run-tests.sh
#!/usr/bin/env sh
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2016 RERO.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, RERO does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.


pydocstyle invenio_collections_metadata tests && \
isort -rc -c -df **/*.py && \
check-manifest --ignore ".travis-*" && \
sphinx-build -qnNW docs docs/_build/html && \
python setup.py test && \
sphinx-build -qnNW -b doctest docs docs/_build/doctest


tests/conftest.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2016 RERO.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, RERO does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.


"""Pytest configuration."""

from __future__ import absolute_import, print_function

import os
import pytest
import shutil
import tempfile

from flask import Flask
from invenio_collections import InvenioCollections
from invenio_collections.models import Collection
from invenio_db import db as db_
from invenio_db import InvenioDB
from invenio_collections_metadata import InvenioCollectionsMetadata
from invenio_collections_metadata.models import CollectionMetadata
from invenio_jsonschemas import InvenioJSONSchemas
from sqlalchemy_utils.functions import create_database, database_exists

SAMPLE_DESCRIPTION = '''<p>
<strong>Lorem ipsum</strong> dolor sit amet, consectetur adipiscing elit. In
quis nunc blandit, imperdiet mi in, blandit dolor. Proin nisi libero, commodo
non tempor et, lacinia at risus. Vestibulum in fermentum lectus. Aenean vel
cursus felis, eu molestie risus. Pellentesque in sagittis eros. Maecenas
porttitor ultrices lectus, et gravida nisi commodo varius. Vestibulum in
tellus interdum, convallis odio ultrices, scelerisque leo. Suspendisse lacinia
massa et convallis fermentum. Sed facilisis justo vel ante dictum elementum.
Nunc rhoncus, justo quis finibus lobortis, dui elit tempus lacus, commodo
porta nunc magna hendrerit dolor. Donec fermentum dui ac ante vestibulum
pharetra. Donec rutrum faucibus convallis. Pellentesque convallis dapibus
eros, quis ultricies lorem elementum in. Quisque laoreet varius mauris sed
tincidunt. Vestibulum sapien dolor, maximus ut interdum sit amet, pretium et
velit.
'''


@pytest.yield_fixture()
def app(request):
    """Flask application fixture."""
    instance_path = tempfile.mkdtemp()
    app_ = Flask(__name__, instance_path=instance_path)
    app_.config.update(
        SECRET_KEY='CHANGE_ME',
        SQLALCHEMY_DATABASE_URI=os.environ.get(
            'SQLALCHEMY_DATABASE_URI', 'sqlite://'),
        SQLALCHEMY_TRACK_MODIFICATIONS=True,
        TESTING=True,
    )
    # app_.register_blueprint(files_rest_blueprint)
    # FlaskCLI(app_)
    InvenioDB(app_)
    InvenioCollections(app_)
    InvenioCollectionsMetadata(app_)
    InvenioJSONSchemas(app_)
    # (app_)

    with app_.app_context():
        yield app_

    shutil.rmtree(instance_path)


@pytest.yield_fixture()
def db(app):
    """Database fixture."""
    if not database_exists(str(db_.engine.url)):
        create_database(str(db_.engine.url))
    db_.create_all()
    yield db_
    db_.session.remove()
    db_.drop_all()


@pytest.fixture()
def collection_metadata(db):
    """Create a bucket."""
    col = Collection(name='test')
    db.session.add(col)
    db.session.commit()
    col_meta = CollectionMetadata(collection=col,
                                  infos={'description': SAMPLE_DESCRIPTION})
    db.session.add(col_meta)
    db.session.commit()
    return col_meta


@pytest.fixture()
def data():
    return {
        'code': 'RERODOC.NAVDOCTYPE.BOOK',
        'parent': 'RERODOC.NAVDOCTYPE',
        'name': {
            'en': 'Book',
            'fr': 'Livres',
            'de': 'Bücher'
        },
        'description': {
            'en': 'A simple <strong>books</strong> collections.',
            'fr': 'Une simple collection de <strong>livres</strong>.'
        },
        # 'thumbnail': {
        #     'url': 'test'
        # }
        'thumbnail': {
            'name': 'book.jpg',
            'bucket': '12345678-1234-1234-1234-123456789012'
        }
    }


tests/test_models.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2016 RERO.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, RERO does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.


"""Module tests."""

from __future__ import absolute_import, print_function

from invenio_collections_metadata.models import CollectionMetadata


def test_data(db, app, collection_metadata):
    assert db
    assert app
    assert collection_metadata
    assert len(collection_metadata.query.all()) == 1
    assert len(CollectionMetadata.query.all()) == 1
    collection = CollectionMetadata.query.first()
    assert collection.collection.name == 'test'
    assert 'test' not in collection.infos.keys()
    assert 'description' in collection.infos.keys()


invenio_collections_metadata/ext.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2016 RERO.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, RERO does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Add metadata to Invenio collections."""

from __future__ import absolute_import, print_function

from flask_babelex import gettext as _

from .views import blueprint


class InvenioCollectionsMetadata(object):
    """Invenio-Collections-Metadata extension."""

    def __init__(self, app=None):
        """Extension initialization."""
        # TODO: This is an example of translation string with comment. Please
        # remove it.
        # NOTE: This is a note to a translator.
        _('A translation string')
        if app:
            self.init_app(app)

    def init_app(self, app):
        """Flask application initialization."""
        self.init_config(app)
        app.register_blueprint(blueprint)
        app.extensions['invenio-collections-metadata'] = self

    def init_config(self, app):
        """Initialize configuration."""
        app.config.setdefault(
            "COLLECTIONS_METADATA_BASE_TEMPLATE",
            app.config.get("BASE_TEMPLATE",
                           "invenio_collections_metadata/base.html"))


invenio_collections_metadata/views.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2016 RERO.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, RERO does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Add metadata to Invenio collections."""

# TODO: This is an example file. Remove it if you do not need it, including
# the templates and static folders as well as the test case.

from __future__ import absolute_import, print_function

from flask import Blueprint, render_template
from flask_babelex import gettext as _

blueprint = Blueprint(
    'invenio_collections_metadata',
    __name__,
    template_folder='templates',
    static_folder='static',
)


@blueprint.route("/")
def index():
    """Basic view."""
    return render_template(
        "invenio_collections_metadata/index.html",
        module_name=_('Invenio-Collections-Metadata'))


invenio_collections_metadata/models.py
"""CollectionMetadata models."""


from invenio_db import db
from invenio_collections.models import Collection
from sqlalchemy.dialects import postgresql
from sqlalchemy_utils.types import JSONType


class CollectionMetadata(db.Model):
    """Represent a collection metadata inside the SQL database.

    """

    __tablename__ = 'collections_metadata'

    """Collection identifier."""
    collection_id = db.Column(
        db.Integer,
        db.ForeignKey(Collection.id),
        primary_key=True,
        nullable=False,
        # NOTE no unique constrain for better future ...
    )

    infos = db.Column(
        JSONType().with_variant(
            postgresql.JSON(none_as_null=True),
            'postgresql',
        ),
        default=lambda: dict(),
        nullable=True
    )

    collection = db.relationship(Collection)


__all__ = (
    'CollectionMetadata',
)


invenio_collections_metadata/version.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2016 RERO.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, RERO does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Version information for Invenio-Collections-Metadata.

This file is imported by ``invenio_collections_metadata.__init__``,
and parsed by ``setup.py``.
"""

from __future__ import absolute_import, print_function

__version__ = "0.1.0.dev20160000"


invenio_collections_metadata/__init__.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2016 RERO.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, RERO does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Add metadata to Invenio collections."""

from __future__ import absolute_import, print_function

from .ext import InvenioCollectionsMetadata
from .version import __version__

__all__ = ('__version__', 'InvenioCollectionsMetadata')


examples/app.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2016 RERO.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, RERO does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.


"""Minimal Flask application example for development.

Run example development server:

.. code-block:: console

   $ cd examples
   $ export FLASK_APP=app.py
   $ export FLASK_DEBUG=1
   $ flask run
"""

from __future__ import absolute_import, print_function

import os

from flask import Flask
from flask_babelex import Babel

from invenio_collections import InvenioCollections
from invenio_collections_metadata import InvenioCollectionsMetadata
from invenio_db import InvenioDB

# Create Flask application
app = Flask(__name__)
app.config.update(
    SECRET_KEY='CHANGE_ME',
    SQLALCHEMY_DATABASE_URI=os.environ.get(
        'SQLALCHEMY_DATABASE_URI', 'sqlite://'),
    SQLALCHEMY_TRACK_MODIFICATIONS=True,
    TESTING=True,
)
Babel(app)
    # instance_path = tempfile.mkdtemp()
InvenioDB(app)
InvenioCollections(application)
InvenioCollectionsMetadata(app)
